mediorum/server/repair.go

```diff
@@ -80,87 +85,109 @@ func (ss *MediorumServer) findProblemBlobsCount(overReplicated bool) (int64, err
 	return count, err
 }
 
-func (ss *MediorumServer) repairUnderReplicatedBlobs() error {
-	// find under-replicated content
+func (ss *MediorumServer) runRepair(cleanupMode bool) error {
 	ctx := context.Background()
-	logger := ss.logger.With("task", "repair")
 
-	problems, err := ss.findProblemBlobs(false)
-	if err != nil {
-		return err
-	}
+	logger := ss.logger.With("task", "repair", "cleanupMode", cleanupMode)
+
+	cidCursor := ""
+	for {
+		// scroll over all extant CIDs in batches
+		// atm this uses `blobs` table for sake of repair_test.go
+		// but if we drop blobs, this could also read out CIDs from the upload records
+		var cidBatch []string
+		err := pgxscan.Select(ctx, ss.pgPool, &cidBatch,
+			`select distinct key
+			 from blobs
+			 where key > $1
+			 order by key
+			 limit 1000`, cidCursor)
 
-	logger.Info("starting repair", "problem_blob_count", len(problems))
+		if err != nil {
+			return err
+		}
+		if len(cidBatch) == 0 {
+			break
+		}
 
-	for _, problem := range problems {
+		for _, cid := range cidBatch {
+			cidCursor = cid
 
-		logger := logger.With("cid", problem.Key)
+			logger := logger.With("cid", cid)
 
-		// if I have the file...
-		// call replicate
-		iHave := strings.Contains(problem.Hosts, ss.Config.Self.Host)
-		if iHave {
-			logger.Info("repairing", "cid", problem.Key)
-			r, err := ss.bucket.NewReader(ctx, problem.Key, nil)
-			if err != nil {
-				logger.Error("failed to read key", err)
+			preferredHosts, isMine := ss.rendezvous(cid)
+
+			// fast path if we're not in cleanup mode:
+			// only worry about blobs that we _should_ have
+			if !cleanupMode && !isMine {
 				continue
 			}
-			defer r.Close()
 
-			// validate CID matches
-			err = validateCID(problem.Key, r)
+			isOnDisk, err := ss.bucket.Exists(ctx, cid)
 			if err != nil {
-				logger.Error("CID validation failed", err)
-				// nuke this fool
-				err = ss.dropFromMyBucket(problem.Key)
-				if err != nil {
-					logger.Error("unable to drop invalid CID", err)
-				}
+				logger.Error("exist check failed", err)
 				continue
 			}
 
-			hosts, err := ss.replicateFile(problem.Key, r)
-			if err != nil {
-				logger.Error("failed to replicate", err)
-			} else {
-				logger.Info("repaired", "hosts", hosts)
+			// todo: in cleanup mode we should do additional check to `isOnDisk`:
+			// todo: validate CID, delete if invalid
+			// todo: check isInDb, create if not
```

Comment:
are these for a followup PR? also if there were an extra `isOnDisk` check wouldn't `ss.bucket.Exists` just return the same bool as before?

Reply:
Yeah that was a comment typo...
but I updated this to do the validate CID check... that is the only one that's actually important.